CLAUDE.md

```diff
@@ -0,0 +1,760 @@
+3. newPayload(executionPayload) → status
+4. forkchoiceUpdate(newHead, safe, finalized) → status
+```
+
+**Key Files:**
+- `execution/client/client.go` - Main Engine API client
+- `execution/client/errors.go` - Error handling and retries
+- `execution/pkg/engine/helpers.go` - Request/response helpers
+
+### State Management Integration
+
+State transitions follow a specific pattern:
+
+```go
+// State transition pipeline
+ctx := state.Context()
+state = processSlots(state, targetSlot)
+state = processBlock(state, block)
+if isEpochEnd(slot) {
+    state = processEpoch(state)
+}
+storage.SetState(ctx, state)
+```
+
+**Context Usage:** All state queries use context for fork-awareness
+
+### Storage Layer Integration
+
+Multi-backend storage with clear interfaces:
+
+```go
+// Storage access pattern
+backend := node.StorageBackend()
+state := backend.StateFromContext(ctx)
+block := backend.BlockStore().GetBySlot(slot)
+blobs := backend.AvailabilityStore().Get(blockRoot)
+deposits := backend.DepositStore().GetAll()
+```
+
+### Service Lifecycle Integration
+
+Services follow a strict lifecycle pattern:
+
+```go
+type Service interface {
+    Start(context.Context) error
+    Stop() error
+    Name() string
+}
+```
+
+**Startup Order:**
+1. Storage backends initialized
+2. Core services created via DI
+3. Services started in dependency order
+4. CometBFT node started last
+
+**Shutdown:** Reverse order with graceful termination
+
+### Configuration Integration
+
+Configuration flows through the system:
+
+```toml
+# config.toml structure
+[beacon-kit]
+  [beacon-kit.engine]
+    jwt-secret-path = "path/to/jwt.hex"
+    rpc-dial-url = "http://localhost:8551"
+
+  [beacon-kit.payload-builder]
+    enabled = true
+
+  [beacon-kit.validator]
+    enable-optimistic-payloads = true
+```
+
+**Environment Overrides:** CLI flags > env vars > config file > defaults
+
+## Block Height Lifecycle
+
+This section describes how BeaconKit processes blocks using CometBFT's Tendermint consensus algorithm.
+
+### Key Differences from Ethereum
+- **Timeout-Based Timing**: No fixed slot duration; block time depends on consensus timeouts and network conditions
+- **Sequential Heights**: CometBFT maintains sequential block heights without gaps (unlike Ethereum's slot system)
+- **Round-Based Consensus**: Multiple proposers per height if rounds fail
+- **Height = Slot**: CometBFT height maps 1:1 to beacon slot number
+
+### Consensus Timing Configuration
+BeaconKit enforces minimum timeout values for CometBFT consensus:
+- **TimeoutPropose**: 2000ms minimum - Initial proposal timeout
+- **TimeoutPrevote**: 2000ms minimum - Prevote collection timeout
+- **TimeoutPrecommit**: 2000ms minimum - Precommit collection timeout
+- **TimeoutCommit**: 500ms minimum - Post-commit wait time
+
+These are **enforced minimums**, not targets. The cumulative timeouts effectively create a minimum block time, but actual block times depend on:
```

Comment:
this is not exactly true.

- The only enforced timeout is timeout_commit.
- The other timeouts are enforced maxima